habitat/tasks/rearrange/rearrange_sensors.py

```diff
@@ -0,0 +1,618 @@
+            high=np.finfo(np.float32).max,
+            dtype=np.float32,
+        )
+
+    def get_observation(self, observations, episode, *args, **kwargs):
+        pos = self._sim.get_target_objs_start()
+        return pos
+
+
+@registry.register_sensor
+class DynObjPosStartOrGoal(MultiObjSensor):
+    """
+    Returns the goal position if the robot is holding an object and the object
+    starting position if the robot is not holding the object.
+    """
+
+    cls_uuid: str = "dyn_obj_start_or_goal_sensor"
+
+    def get_observation(self, observations, episode, *args, **kwargs):
+        global_T = self._sim.robot.ee_transform
+        T_inv = global_T.inverted()
+
+        if self._sim.snapped_obj_id is not None:
+            _, pos = self._sim.get_targets()
+        else:
+            pos = self._sim.get_target_objs_start()
+        for i in range(pos.shape[0]):
+            pos[i] = T_inv.transform_point(pos[i])
+
+        return pos
+
+
+@registry.register_sensor
+class GoalSensor(MultiObjSensor):
+    """
+    Relative to the end effector
+    """
+
+    cls_uuid: str = "obj_goal_sensor"
+
+    def get_observation(self, observations, episode, *args, **kwargs):
+        global_T = self._sim.robot.ee_transform
+        T_inv = global_T.inverted()
+
+        _, pos = self._sim.get_targets()
+        for i in range(pos.shape[0]):
+            pos[i] = T_inv.transform_point(pos[i])
+        return pos
+
+
+@registry.register_sensor
+class AbsGoalSensor(MultiObjSensor):
+    cls_uuid: str = "abs_obj_goal_sensor"
+
+    def get_observation(self, observations, episode, *args, **kwargs):
+        _, pos = self._sim.get_targets()
+        return pos
+
+
+@registry.register_sensor
+class DummySensor(Sensor):
+    def _get_uuid(self, *args, **kwargs):
+        return "dummy"
+
+    def _get_sensor_type(self, *args, **kwargs):
+        return SensorTypes.TENSOR
+
+    def _get_observation_space(self, *args, **kwargs):
+        return spaces.Box(
+            shape=(1,),
+            low=np.finfo(np.float32).min,
+            high=np.finfo(np.float32).max,
+            dtype=np.float32,
+        )
+
+    def get_observation(self, observations, episode, *args, **kwargs):
+        return np.zeros(1)
+
+
+@registry.register_sensor
+class LocalizationSensor(Sensor):
+    def __init__(self, sim, config, *args, **kwargs):
+        super().__init__(config=config)
+        self._sim = sim
+
+    def _get_uuid(self, *args, **kwargs):
+        return "localization"
+
+    def _get_sensor_type(self, *args, **kwargs):
+        return SensorTypes.TENSOR
+
+    def _get_observation_space(self, *args, **kwargs):
+        return spaces.Box(
+            shape=(4,),
+            low=np.finfo(np.float32).min,
+            high=np.finfo(np.float32).max,
+            dtype=np.float32,
+        )
+
+    def get_observation(self, observations, episode, *args, **kwargs):
+        trans = self._sim.robot.sim_obj.transformation
+        forward = np.array([1.0, 0, 0])
```

Comment:
The coordinate frame in habitat is that `-Z` is forward, why is this being done differently here?

Reply:
@ASzot

Reply:
There's also another instance of forward being defined like this in `hab_simulator.py`

Reply:
Yes, I think I had changed the coordinate system in a config file in my experiments, however, this is not correct in this branch. I am working on fixing this.